distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso

```diff
@@ -45,6 +47,241 @@ type Dialect_Operations
     is_supported self operation =
         self.operations_dict.contains_key operation
 
+## Private
+
+   Generates SQL from a `Query`.
+
+   SQL generation requires some state threaded through the generation process,
+   and that state is stored in the fields of this type. This type only contains
+   the mutually recursive generators for the SQL IR that require this state.
+type SQL_Generator
+    ## We can use a Hashset for the let-binding names, since they are guaranteed
+       to be distinct and not result in any shadowing.
+    Value (let_bindings_ref : Ref Hashset)
```

Comment:
I'd like to understand the design decision here - why did you prefer to use `Ref` for this instead of `State`?

Reply:
I assumed that `Ref` is preferred, since it's used much more in our codebase than `State`. It also means no one will be tempted to mess with the bindings from anywhere else, since it has to be accessed from inside this class. So it seems slightly less imperative in style. I'm happy to revert it to State (and remove the class) if that's better.

Reply:
Hmm, good point. Let's keep it that way for now then.